File: 040_Isomorphic Strings.cpp
/*
Isomorphic Strings

Given two strings s and t, determine if they are isomorphic. Characters must map one-to-one.
*/

#include <bits/stdc++.h>
using namespace std;

bool isIsomorphic(string s, string t) {
    if (s.size() != t.size()) return false;
    vector<int> m1(256, -1), m2(256, -1);
    for (int i = 0; i < (int)s.size(); ++i) {
        unsigned char a = s[i], b = t[i];
        if (m1[a] != m2[b]) return false;
        m1[a] = m2[b] = i;
    }
    return true;
}

int main() {
    cout << "Iso 1: " << (isIsomorphic("egg","add")?"true":"false") << endl; // true
    cout << "Iso 2: " << (isIsomorphic("foo","bar")?"true":"false") << endl; // false
    cout << "Iso 3: " << (isIsomorphic("paper","title")?"true":"false") << endl; // true
    return 0;
}

// Time Complexity: O(n)
// Space Complexity: O(1) (fixed alphabet)

File: 040_Isomorphic Strings.py
"""
Isomorphic Strings

Given two strings s and t, determine if they are isomorphic. Characters must map one-to-one.
"""

def is_isomorphic(s: str, t: str) -> bool:
    if len(s) != len(t): return False
    m1, m2 = {}, {}
    for i, ch in enumerate(s):
        if ch in m1 and m1[ch] != t[i]: return False
        if t[i] in m2 and m2[t[i]] != ch: return False
        m1[ch] = t[i]
        m2[t[i]] = ch
    return True


if __name__ == '__main__':
    print('Iso 1:', is_isomorphic('egg','add'))
    print('Iso 2:', is_isomorphic('foo','bar'))
    print('Iso 3:', is_isomorphic('paper','title'))

# Time Complexity: O(n)
# Space Complexity: O(min(n, charset))
